### ISSUE TITLE
Is it possible to change value of num_multimask_outputs to greater than 3?

### ISSUE THREAD

**Issue by DavidLanders95** [facebookresearch/segment-anything#299]
Is is possible to change the number of possible mask outputs from the model to another value greater than 3? I see in the class MaskDecoder there is an option: 

```python
class MaskDecoder(nn.Module):

    def __init__(
        self,
        *,
        transformer_dim: int,
        transformer: nn.Module,
        num_multimask_outputs: int = 3,
        activation: Type[nn.Module] = nn.GELU,
        iou_head_depth: int = 3,
        iou_head_hidden_dim: int = 256,
    ) 
```

However, when I try and edit this value to say 5 or 10, I receive an error: 

> RuntimeError: Error(s) in loading state_dict for Sam:
> 	Missing key(s) in state_dict: "mask_decoder.output_hypernetworks_mlps.4.layers.0.weight", "mask_decoder.output_hypernetworks_mlps.4.layers.0.bias", "mask_decoder.output_hypernetworks_mlps.4.layers.1.weight", "mask_decoder.output_hypernetworks_mlps.4.layers.1.bias", "mask_decoder.output_hypernetworks_mlps.4.layers.2.weight", "mask_decoder.output_hypernetworks_mlps.4.layers.2.bias", "mask_decoder.output_hypernetworks_mlps.5.layers.0.weight", "mask_decoder.output_hypernetworks_mlps.5.layers.0.bias", "mask_decoder.output_hypernetworks_mlps.5.layers.1.weight", "mask_decoder.output_hypernetworks_mlps.5.layers.1.bias", "mask_decoder.output_hypernetworks_mlps.5.layers.2.weight", "mask_decoder.output_hypernetworks_mlps.5.layers.2.bias". 
> 	size mismatch for mask_decoder.mask_tokens.weight: copying a param with shape torch.Size([4, 256]) from checkpoint, the shape in current model is torch.Size([6, 256]).
> 	size mismatch for mask_decoder.iou_prediction_head.layers.2.weight: copying a param with shape torch.Size([4, 256]) from checkpoint, the shape in current model is torch.Size([6, 256]).
> 	size mismatch for mask_decoder.iou_prediction_head.layers.2.bias: copying a param with shape torch.Size([4]) from checkpoint, the shape in current model is torch.Size([6]).

Does anyone know if this is possible to solve?

**Comment by starsky0426** [comment]
Yes you can change it, but you need to train the network again. 😅

**Comment by EveningLin** [comment]
`self.num_mask_tokens` 
why is 4？

**Comment by DavidLanders95** [comment]
> 



> Yes you can change it, but you need to train the network again. 😅

Thanks, I thought that might be the case.



> `self.num_mask_tokens` why is 4？

I'm not sure why, it's one more than the number of multimask outputs for some reason. It's set in the code.

**Comment by halleewong** [comment]
> `self.num_mask_tokens` why is 4？

When `multimask_output` is False, the predicted mask associated with the first mask token is returned. Otherwise, the masks associated with the other 3 masks tokens are returned.

**Comment by Soulergonote** [comment]
Is there a way to generate more than 3 masks for one input whithout retraining the network ?

**Comment by Paul-Guo** [comment]
@DavidLanders95 You can see my solution. [https://github.com/facebookresearch/sam2/issues/439#issuecomment-2709623412]

**Comment by 2384441037** [comment]
> [@DavidLanders95](https://github.com/DavidLanders95)你可以查看我的解决方案。[https://github.com/facebookresearch/sam2/issues/439 [#](https://github.com/facebookresearch/sam2/issues/439) issuecomment-2709623412]

请问一下，num_multimask_outputs这个参数是用来控制具体分类的种类数吗，我想实现多种类分割并且multimask_output设置为true，mask返回是对应种类的分割结果吗
